res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,772 @@
+
+    // initialize leds for both decks
+    for (var i = 0; i < 2; i++) {
+        midi.sendShortMsg(0x90 + i, 0x00, 0x01); // play
+        midi.sendShortMsg(0x90 + i, 0x01, 0x01); // cue
+        midi.sendShortMsg(0x90 + i, 0x02, 0x01); // sync
+        midi.sendShortMsg(0x90 + i, 0x07, 0x7f); // scratch
+        midi.sendShortMsg(0x90 + i, 0x1B, 0x01); // pfl
+
+        midi.sendShortMsg(0x94 + i, 0x00, 0x7F); // hotcue
+        midi.sendShortMsg(0x94 + i, 0x0D, 0x01); // auto loop
+        midi.sendShortMsg(0x94 + i, 0x07, 0x01); // fader cuts
+        midi.sendShortMsg(0x94 + i, 0x0B, 0x01); // sample
+
+        midi.sendShortMsg(0x94 + i, 0x34, 0x01); // half
+        midi.sendShortMsg(0x94 + i, 0x35, 0x01); // double
+        midi.sendShortMsg(0x94 + i, 0x40, 0x01); // loop
+
+        // initialize leds for shifted buttons
+        midi.sendShortMsg(0x90 + i, 0x04, 0x01); // play
+        midi.sendShortMsg(0x90 + i, 0x05, 0x01); // cue
+        midi.sendShortMsg(0x90 + i, 0x03, 0x01); // sync
+        midi.sendShortMsg(0x94 + i, 0x0F, 0x01); // sample shifted
+        midi.sendShortMsg(0x94 + i, 0x02, 0x01); // beatjump
+        midi.sendShortMsg(0x90 + i, 0x08, 0x01); // bleep
+        midi.sendShortMsg(0x94 + i, 0x36, 0x01); // half
+        midi.sendShortMsg(0x94 + i, 0x37, 0x01); // double
+        midi.sendShortMsg(0x94 + i, 0x41, 0x01); // loop
+
+        // pads
+        for (var j = 0; j < 8; j++) {
+            midi.sendShortMsg(0x94 + i, 0x14 + j, 0x01);
+            midi.sendShortMsg(0x94 + i, 0x1C + j, 0x01); // shifted
+        }
+    }
+
+    midi.sendShortMsg(0x88, 0x09, 0x01); // tap led
+
+    // check if quantize is enabled
+    var quantizeEnabled = engine.getValue("[Channel1]", "quantize");
+
+    // effect leds
+    midi.sendShortMsg(0x88, 0x00, 0x01); // hpf
+    midi.sendShortMsg(0x88, 0x01, 0x01); // lpf
+    midi.sendShortMsg(0x88, 0x02, 0x01); // flanger
+    midi.sendShortMsg(0x89, 0x03, 0x01); // echo
+    midi.sendShortMsg(0x89, 0x04, quantizeEnabled ? 0x7F : 0x01); // reverb
+    midi.sendShortMsg(0x89, 0x05, 0x01); // phaser
+
+    // vumeters leds (off)
+    midi.sendShortMsg(0xB0, 0x1F, 0x00);
+    midi.sendShortMsg(0xB1, 0x1F, 0x00);
+
+    engine.makeConnection("[Channel1]", "VuMeter", MixtrackProFX.vuCallback);
+    engine.makeConnection("[Channel2]", "VuMeter", MixtrackProFX.vuCallback);
+};
+
+MixtrackProFX.shutdown = function() {
+    var shutdownSysex = [0xF0, 0x00, 0x20, 0x7F, 0x02, 0xF7];
+    midi.sendSysexMsg(shutdownSysex, shutdownSysex.length);
+};
+
+MixtrackProFX.EffectUnit = function(unitNumber) {
+    this.unitNumber = unitNumber;
+    this.group = "[EffectRack1_EffectUnit" + unitNumber + "]";
+
+    this.enableButton = new components.Button({
+        input: function(channel, control, value, status, group) {
+            if (value === 2) {
+                value = 1;
+            }
+            engine.setValue(group, "enabled", value);
```

Comment:
The switch that enables FX has two positions - up (held in place), and down (automatically jumps back). The controller sends 1 when the switch is in the "up" position, 2 when in "down". I just checked and it looks like the engine interprets both 1 and 2 as "enabled", so I will remove this check. I think I was testing it on older Mixxx version and conversion to 1 was needed then.

Reply:
I wouldn't rely on that. Instead of overwriting each input method, just create a single function that takes care of the clamping, and then set that as the inValueScale method on the components that need this.

Reply:
Does the controller send 0 when the switch automatically goes back to center after pushing it down? Please document this information in a comment.

Reply:
Yes it sends 0 in the middle position. I've added a comment.

I'm having trouble with inValueScale - I think the Button component is ignoring this property. Am I doing it right?:
```
MixtrackProFX.scaleToBinaryInput = function(value) {
    if (value > 0) {
        return 1;
    }
    return 0;
};

// switch values are:
// 0 - switch in the middle
// 1 - switch up
// 2 - switch down
this.enableSwitch = new components.Button({
    group: "[EffectRack1_EffectUnit" + deckNumber + "_Effect1]",
    inKey: "enabled",
    inValueScale: MixtrackProFX.scaleToBinaryOutput
 });
```

Reply:
That code looks correct. Were you able to confirm that `scaleToBinaryInput` actually gets called? you can use `print()` for console-based debugging.

Reply:
`scaleToBinaryInput` is not getting called when I change the switch state. When I set `inValueScale` to some invalid value, like:
```
inValueScale: MixtrackProFX.asdfasdf
```
I'm not even getting an error. I'm using Mixxx 2.3.0

Reply:
Can you commit and push the broken code so I can look at it in context?

Reply:
> I'm not even getting an error. I'm using Mixxx 2.3.0

That might be because the javascript engine used in 2.3.0 just silently fails in some obscure cases. Can you try the 2.4 alpha and see if works then?

Reply:
`2.4-alpha-735-gd06bc0edd7 (main)` works the same for me (`inValueScale` function is not getting called).

Furthermore I'm getting an initialization error in 2.4. I don't understand what it means - I don't use `outValueScale` anywhere in my code.

```
Uncaught exception: file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:73: TypeError: Property 'outValueScale' of object [object Object] is not a function
Backtrace: output@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:73
@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:139
trigger@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:138
@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:485
applyOperationTo@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:462
forEachComponent@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:474
reconnectComponents@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/midi-components-0.0.js:479
@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/Numark-Mixtrack-Pro-FX-scripts.js:386
@file:///C:/Users/sancho/AppData/Local/Mixxx/controllers/Numark-Mixtrack-Pro-FX-scripts.js:84
```

I've pushed the current code. Functionally it works fine (in 2.3), just the `inValueScale` is not getting called.

Reply:
well `outValueScale` is used by default `output` implementation. The default Button implementation provides an implementation that automatically sends the `on` value when the CO value is greater than 0. However, for some obscure reason, this seems to be `undefined` in your case which is why it throws the `TypeError`. I'll be trying to reproduce this, though unfortunately my time is quite limited the next week.

Reply:
I just reread the components code and remembered that in case of the button, we actually don't use inValueScale (since the button was designed with only an off or on state in mind, `inValueScale` would be slightly confusing). Instead you would want to provide a custom `isPress: function(channel, control, value, _status)` implementation (which returns a boolean). However, the default isPress method should just work out of the box. So there should not have to be any need to customize the enableSwitch at all.

Reply:
Thank you for verifying this, so in the end value scaling is not needed here.